toolbox/src/metaplex_core.rs

```diff
@@ -17,6 +22,92 @@ impl anchor_lang::Id for MetaplexCore {
+    pub royalty_enforced: bool,
+}
+
+/// Validates a mpl-core asset.
+///
+/// Validates program ownership and if it is part of a collection and extracts royalty and verified creators information
+/// from the appropriate plugins.
+pub fn validate_core_asset(
+    asset_info: &AccountInfo,
+    maybe_collection_info: Option<&AccountInfo>,
+) -> Result<CoreAsset> {
+    // validate the asset account
+    assert_ownership(asset_info, Key::AssetV1)?;
+
+    // validates the collection is owned by the MPL Core program
+    maybe_collection_info
+        .as_ref()
+        .map(|c| assert_ownership(c, Key::CollectionV1));
+
+    let asset = BaseAssetV1::try_from(asset_info)?;
+
+    // if the asset has a collection, we must validate it, and fetch the royalties from it
+    let (mut royalties, collection) =
+        if let UpdateAuthority::Collection(asset_collection) = asset.update_authority {
+            // Collection account must be provided.
+            if maybe_collection_info.is_none() {
+                msg!("Asset has a collection but no collection account was provided");
+                return Err(TensorError::InvalidCoreAsset.into());
+            }
+
+            let collection_info = maybe_collection_info.unwrap();
+
+            // Collection account must match the one on the asset.
+            if asset_collection != *collection_info.key {
+                msg!("Asset collection account does not match the provided collection account");
+                return Err(TensorError::InvalidCoreAsset.into());
+            }
+
+            (
+                fetch_plugin::<BaseCollectionV1, Royalties>(collection_info, PluginType::Royalties)
+                    .ok()
+                    .map(|(_, royalties, _)| royalties),
+                Some(asset_collection),
+            )
+        } else {
+            (None, None)
+        };
+
+    // Get royalties plugin from the asset, if present.
+    // Royalties plugin directly on the asset take precedence over the one on the collection.
+    royalties = fetch_plugin::<BaseAssetV1, Royalties>(asset_info, PluginType::Royalties)
+        .ok()
+        .map(|(_, royalties, _)| royalties)
+        .or(royalties);
+
+    let royalty_fee_bps = if let Some(Royalties { basis_points, .. }) = royalties {
+        basis_points
+    } else {
+        0
+    };
+
+    // Fetch the verified creators from the MPL Core asset and map into the expected type
+    // for whitelist verification.
+    let verified_creators: Option<Vec<VerifiedCreatorsSignature>> =
+        fetch_plugin::<BaseAssetV1, VerifiedCreators>(asset_info, PluginType::VerifiedCreators)
+            .map(|(_, verified_creators, _)| verified_creators.signatures)
+            .ok();
+
+    Ok(CoreAsset {
+        pubkey: *asset_info.key,
+        collection,
+        whitelist_creators: verified_creators,
+        royalty_creators: royalties.map(|r| r.creators),
+        royalty_fee_bps,
+        royalty_enforced: true,
+    })
+}
```

Comment:
So this will replace `validate_asset` right? If so, remember to remove `validate_asset` from all uses + toolbox goforward.

Reply:
Yeah, I've removed it from toolbox now and have it removed from AMM in an open PR.